plugin-infra/go-plugin-infra/src/com/thoughtworks/go/plugin/infra/monitor/DefaultPluginJarLocationMonitor.java

```diff
@@ -181,40 +183,50 @@ public void run() {
 
         private Set<PluginFileDetails> loadAndNotifyPluginsFrom(File pluginDirectory, Set<PluginFileDetails> knownPluginFiles, boolean isBundledPluginsLocation) {
             Set<PluginFileDetails> currentPluginFiles = getDetailsOfCurrentPluginFilesFrom(pluginDirectory, isBundledPluginsLocation);
-            notifyListenersOfRemovedPlugins(currentPluginFiles, knownPluginFiles);
-            notifyListenersOfUpdatedPlugins(currentPluginFiles, knownPluginFiles);
-            notifyListenersOfAddedPlugins(currentPluginFiles, knownPluginFiles);
+            boolean notifyListenersOfRemovedPlugins = notifyListenersOfRemovedPlugins(currentPluginFiles, knownPluginFiles);
+            boolean notifyListenersOfUpdatedPlugins = notifyListenersOfUpdatedPlugins(currentPluginFiles, knownPluginFiles);
+            boolean notifyListenersOfAddedPlugins = notifyListenersOfAddedPlugins(currentPluginFiles, knownPluginFiles);
+            if (notifyListenersOfRemovedPlugins || notifyListenersOfUpdatedPlugins || notifyListenersOfAddedPlugins) {
```

Comment:
No.
While it looks like a simple check similar to *currentPluginFiles!=knownPluginFiles* will work, it won't!

We need the three boolean values from the three methods since the current and known plugin files might have same name but different md5 (plugin update case). These methods take care of the md5 changes while comparing the plugin files and it would be duplication and unnecessary work to do that again over here. I thought it better to check on the boolean than do the check again.